### PR TITLE
[infrastructure] embed single-use-dependencies 

### PR DESCRIPTION
**Pull request by J-N-K**
This PR enables embedding of dependencies which are used only by one bundle by unpacking the .jar files and embedding the classes in the final bundle.

Also changes all scope "provided" dependencies to scope "compile" dependencies and adds exceptions for dependencies that are used in multiple bundles.  

This should solve at least most of the bnd resolution problems in the demo app. 

Some bundles that have single-use dependencies (especially with native code) look a little bitmore difficult and should be adapted in later PRs, they are excluded here,  

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

**Comment by martinvw** [comment]
@J-N-K interesting errors:

> [INFO] --- bnd-maven-plugin:4.2.0:bnd-process (default) @ org.openhab.addons.bom.test-index ---
[WARNING] /home/travis/build/openhab/openhab2-addons/bom/test-index/pom.xml [0:0]: Unused Export-Package instructions: [org.openhab.*] 
[WARNING] /home/travis/build/openhab/openhab2-addons/bom/test-index/pom.xml [0:0]: Unused Import-Package instructions: [io.swagger.annotations.*, javax.annotation.security.*, org.eclipse.jdt.annotation.*, org.openhab.core.automation.annotation.*, org.eclipse.smarthome.*, org.openhab.*, com.google.common.*] 
[ERROR] /home/travis/build/openhab/openhab2-addons/bom/test-index/pom.xml [0:0]: Classes found in the wrong directory: {META-INF/versions/9/module-info.class=module-info}

**Comment by J-N-K** [comment]
This can easily be fixed by movinf the plugin-section to the bundles-pom.

**Comment by J-N-K** [comment]
I have no idea what Jenkins is talking about. I prefer this way, there is only one questions which i can‘t answer: if we unpack and embed the dependencies on class-level and add the original contribution and license to the NOTICE file, can we run into IP issues? 

With the approach here we can also do two other things: we can just upload the libraries in /lib and they will get embedded like before, without wrapping or other difficult stuff and remove them from our source-repo and second: with another maven plugin we can probably only include those classes we really need and reduce the download size. WDYT?

**Comment by J-N-K** [comment]
This would be my proposal for dealing with dependencies. 

I prepared some "example bindings": 

- allplay (native code)
- amazondashbutton (partly embedded, not working without embedding)
- bluetooth.bluez (native code)
- mail
- snmp

**Benefits:**

- No resolution problems if the bundle is used by one binding only, they are internal to our bundle. 
- Only classes that are really used are imported, greatly reduced download size (e.g. 1488 kB vs. 132 kB for amazondashbutton)
- No need to wrap bundles, all dependencies that are not shared between bindings can be used as-is, no matter if they are OSGi-bundles or not.
- No problem with native code, it stays within the bundle. 
- Simpler features, dependencies only need to be added if they are shared between bindings.

**Features:**

- by default all `compile`-scope dependencies are embedded, all unused classes are removed, all classes in `org.openhab.*` classes are always kept
- the `dep.noembedding`-property can be used to exclude bundles from embedding by their artifactId
- the `no-embed-dependencies`-profile (activated by adding a zero-length `noEmbedDependencies.profile` file) prevents embedding dependencies
- the `no-shrink-bundle`-profile (activated by adding a zero-length `noShrinkBundle.profile` file) prevents execution of the shrinker

@openhab/2-x-add-ons-maintainers, WDYT?

**Comment by kaikreuzer** [comment]
Thanks for your work, @J-N-K, this sounds pretty good to me!

> if we unpack and embed the dependencies on class-level and add the original contribution and license to the NOTICE file, can we run into IP issues?

No, I think if we correctly mention the libs in the NOTICE file (the same way we did when including them in the lib folder), this should be all perfectly fine from the IP perspective.

> we can just upload the libraries in /lib and they will get embedded like before

I think we should avoid this - binaries do not belong into the source code repo and we only did it that way for Tycho because we had no other way to easily reference them. But as we can now simply refer to them through Maven dependencies, that's imho the much cleaner approach.

> Only classes that are really used are imported, greatly reduced download size

This is nice, but it also sounds as if there's a risk that too much is removed in case that classes are accessed through reflection at runtime. So when using this feature, we should carefully test that there's no issue at runtime for the binding.

File: bundles/pom.xml [hunk 1]
```diff
@@ -364,7 +369,106 @@
...
+            <executions>
+              <execution>
+                <id>embed-dependencies</id>
+                <phase>none</phase>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile
```
Comment (kaikreuzer) [inline bundles/pom.xml]:
```suggestion
    </profile>
```

**Comment by J-N-K** [comment]
> > we can just upload the libraries in /lib and they will get embedded like before
> 
> I think we should avoid this - binaries do not belong into the source code repo and we only did it that way for Tycho because we had no other way to easily reference them. But as we can now simply refer to them through Maven dependencies, that's imho the much cleaner approach.

I should rephrase that. I meant: We can upload all those files we currently have in /lib to one of our bintray-repos and if we add them as dependency to the binding's pom.xml they'll get embedded the same way as before. We can fully clean our source-repo from binaries then.

> > Only classes that are really used are imported, greatly reduced download size
> This is nice, but it also sounds as if there's a risk that too much is removed in case that classes are accessed through reflection at runtime. So when using this feature, we should carefully test that there's no issue at runtime for the binding.

Yeah. Just found the first issue with that. Maybe we should turn the activation around: de-activated by default and can be enabled for bindings where we are sure it works.

**Comment by J-N-K** [comment]
I would like to document how to use the new embedding features for dependencies. What would be the best place to do that?

**Comment by J-N-K** [comment]
@openhab/2-x-add-ons-maintainers could you review this before we get merge conflicts?

**Comment by wborn** [comment]
Perhaps these changes caused the Jenkins build to fail?

https://ci.openhab.org/job/openHAB2-Bundles/771/console

**Comment by J-N-K** [comment]
Maybe some caching issues? org.openhab.io.javasound should not import javazoom.* anymore, it‘s embedded.

**Comment by wborn** [comment]
Looks like a parallel Maven build was configured again (-T 1C) so I've removed that again.

**Pull request merged**